src/types.rs

```diff
@@ -438,6 +466,153 @@ impl PartialEq for Highlight {
+                }
+            }
+        }
+    }
+
+    fn slices_match(a: &[String], b: &[String]) -> bool {
+        if a.len() == b.len() {
+            for (i, v) in a.iter().enumerate() {
+                if *v != "*" && b[i] != *v {
+                    return false;
+                }
+            }
+
+            true
+        } else {
+            false
+        }
+    }
+}
+
+#[test]
+fn test_semantic_hl_matcher_str() {
+    let matcher = SemanticHighlightMatcher::Str("Hello".into());
+
+    assert!(matcher.matches(&vec!["Hello".into()]));
+    assert!(matcher.matches(&vec![
+        "X".into(),
+        "HELLO".into(),
+        "Hello".into(),
+        "ABCD".into()
+    ]));
+    assert!(matcher.matches(&vec!["Hello".into(), "ABCD".into(), "X".into()]));
+    assert!(!matcher.matches(&vec!["ABCD".into(), "X".into()]));
+}
+
+#[test]
+fn test_semantic_hl_matcher_array() {
+    let v = vec!["A".into(), "B".into()];
+    let arr = SemanticHighlightMatcher::Array(v.clone());
+    let arr_s = SemanticHighlightMatcher::ArrayStart(v.clone());
+    let arr_e = SemanticHighlightMatcher::ArrayEnd(v.clone());
+    let arr_c = SemanticHighlightMatcher::ArrayContains(v.clone());
+
+    let t1 = vec!["A".into(), "B".into()];
+    let t2 = vec!["FFF".into(), "A".into(), "B".into(), "ABCD".into()];
+    let t3 = vec!["Hello".into(), "A".into(), "B".into()];
+    let t4 = vec!["A".into(), "B".into(), "CC".into(), "D".into()];
+
+    let do_matches = |v: &[String]| -> (bool, bool, bool, bool) {
+        (
+            arr.matches(v),
+            arr_s.matches(v),
+            arr_e.matches(v),
+            arr_c.matches(v),
+        )
+    };
+
+    assert_eq!(do_matches(&t1), (true, true, true, true));
+    assert_eq!(do_matches(&t2), (false, false, false, true));
+    assert_eq!(do_matches(&t3), (false, false, true, true));
+    assert_eq!(do_matches(&t4), (false, true, false, true));
+}
+
+#[test]
+fn test_semantic_hl_matcher_array_glob() {
+    let v = vec!["A".into(), "*".into()];
+    let arr = SemanticHighlightMatcher::Array(v.clone());
+    let arr_s = SemanticHighlightMatcher::ArrayStart(v.clone());
+    let arr_e = SemanticHighlightMatcher::ArrayEnd(v.clone());
+    let arr_c = SemanticHighlightMatcher::ArrayContains(v.clone());
+
+    let t1 = vec!["A".into(), "B".into()];
+    let t2 = vec!["FFF".into(), "A".into(), "B".into(), "ABCD".into()];
+    let t3 = vec!["Hello".into(), "A".into(), "B".into()];
+    let t4 = vec!["A".into(), "B".into(), "CC".into(), "D".into()];
+
+    let do_matches = |v: &[String]| -> (bool, bool, bool, bool) {
+        (
+            arr.matches(v),
+            arr_s.matches(v),
+            arr_e.matches(v),
+            arr_c.matches(v),
+        )
+    };
+
+    assert_eq!(do_matches(&t1), (true, true, true, true));
+    assert_eq!(do_matches(&t2), (false, false, false, true));
+    assert_eq!(do_matches(&t3), (false, false, true, true));
+    assert_eq!(do_matches(&t4), (false, true, false, true));
+}
+
```

Comment:
Can you add unit tests covering case of `**`?

Reply:
When a `**` gets used it gets converted into the corresponding `SemanticHighlightMatcher` so the string `**` doesn't ever appear in the matcher itself. See `buildSemanticHighlightMatchers` in `language_server_protocol.rs` for the construction.

E.g. an array `['a', 'b', 'c', '**']` would construct the `ArrayStart` matcher.

This is kind of limiting right now since the `**` can only be used at the beginning/end of the array (although it should be sufficient for most use cases). I don't see any good way to evaluate arbitrary patterns of `**` without an overly complex implemenation or hacks like concatentating the items and regexing. Any suggestions?